tdrs-backend/tdpservice/data_files/error_reports.py

```diff
@@ -0,0 +1,341 @@
+        """To be overriden in child class."""
+        pass
+
+    @abstractmethod
+    def get_row_generator(self):
+        """Get row generator for error report."""
+        pass
+
+    @abstractmethod
+    def get_columns(self):
+        """Get the columns for header."""
+        pass
+
+    def format_header(self, header_list: list):
+        """Format header."""
+        return ' '.join([i.capitalize() for i in header_list.split('_')])
+
+    def format_error_msg(self, error_msg, fields_json):
+        """Format error message."""
+        for key, value in fields_json['friendly_name'].items():
+            error_msg = error_msg.replace(key, value) if value else error_msg
+        return error_msg
+
+    def friendly_names(self, fields_json):
+        """Return comma separated string of friendly names."""
+        return ','.join([i for i in fields_json['friendly_name'].values()])
+
+    def internal_names(self, fields_json):
+        """Return comma separated string of internal names."""
+        return ','.join([i for i in fields_json['friendly_name'].keys()])
+
+    def check_fields_json(self, fields_json, field_name):
+        """If fields_json is None, impute field name to avoid NoneType errors."""
+        if not fields_json:
+            child_dict = {field_name: field_name} if field_name else {}
+            fields_json = {'friendly_name': child_dict}
+        return fields_json
+
+
+class FRADataErrorReport(ErrorReportBase):
+    """FRA Error Report generator."""
+
+    def __init__(self, datafile):
+        super().__init__(datafile)
+
+    def generate(self):
+        """Generate and return FRA error report."""
+        worksheet = self.workbook.add_worksheet(name="Error Report")
+
+        bold = self.workbook.add_format({'bold': True})
+        columns = self.get_columns()
+        for idx, col in enumerate(columns):
+            worksheet.write(0, idx, self.format_header(col), bold)
+
+        row_idx = 1
+        for error in self.parser_errors.iterator():
+            ssn = error.values_json.get('SSN', '')
+            exit_date = error.values_json.get('EXIT_DATE', None)
+            fields_json = self.check_fields_json(getattr(error, 'fields_json', {}), error.field_name)
+            row = self.row_generator(error, exit_date, ssn, fields_json)
+            worksheet.write_row(row_idx, 0, row)
+            row_idx += 1
+
+        self.workbook.close()
+        return {"xls_report": base64.b64encode(self.output.getvalue())}
+
+    def get_columns(self):
+        """Get the columns for header."""
+        return ["exit_date", "ssn", "error_location_in_file", "error_description"]
+
+    def _obscure_ssn(self, ssn):
+        """Obscure SSN."""
+        return f"*****{ssn[-4:]}"
+
+    def get_row_generator(self):
+        """Get row generator for error report."""
+        return lambda error, exit_date, ssn, fields_json: (exit_date,
+                                                           self._obscure_ssn(ssn),
```

Comment:
this (and the default `''` for `ssn`) results in the cat1/header error `File does not begin with FRA data` to also have a partial obscured ssn. suggest defaulting to `None` and `self._obscure_ssn(ssn) if ssn else ''`